scipy/stats/_hypotests.py

```diff
@@ -664,34 +663,45 @@ def somersd(x, y=None):
         raise ValueError("x must be either a 1D or 2D array")
     d, p = _somers_d(table)
     return SomersDResult(d, p, table)
+
+
+def _all_partitions(nx, ny):
+    """
+    Partition a set of indices into two fixed-length sets in all possible ways
+
+    Partition a set of indices 0 ... nx + ny - 1 into two sets of length nx and
+    ny in all possible ways (ignoring order of elements).
+    """
+    z = np.arange(nx+ny)
+    for c in combinations(z, nx):
+        x = np.array(c)
+        mask = np.ones(nx+ny, bool)
+        mask[x] = False
+        y = z[mask]
+        yield x, y
```

Comment:
I was suggesting this change not for the little improvement, it is because I think making a difference with sets add readability comparing to using masks (but it is only my opinion). The improvement is only a bonus.

Reply:
Sounds like we're going to replace this with Pythran anyway, so I'll leave it alone for now.